File: Ghostpad-native/src/network/payload_deploy.cpp
// Ghostpad Native - PS5 Remote Controller
// Copyright (c) 2026  seregonwar
// Based on original Ghostpad by stonedmodder  
// Licensed under the GNU General Public License v3.0. See LICENSE file for details.

#include "network/payload_deploy.h"
#include "protocol/gpad_packet.h"
#include <cstring>
#include <fstream>
#include <regex>
#include <thread>
#include <arpa/inet.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#else
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#endif

namespace ghostpad {

static void closeSocket(int sock) {
    if (sock < 0) return;
#ifdef _WIN32
    closesocket(sock);
#else
    close(sock);
#endif
}

static void setNoSigPipe(int sock) {
#ifdef __APPLE__
    int opt = 1;
    setsockopt(sock, SOL_SOCKET, SO_NOSIGPIPE, &opt, sizeof(opt));
#elif !defined(_WIN32)
    (void)sock;
#endif
}

PayloadDeployer::PayloadDeployer() {
    status_.phase = "idle";
    status_.message = "Ready";
}

PayloadDeployer::~PayloadDeployer() {
    stopKlogWatcher();
}

void PayloadDeployer::emitStatus(const std::string& phase, const std::string& message) {
    std::lock_guard<std::mutex> lock(mutex_);
    status_.phase = phase;
    status_.message = message;
    status_.at = std::to_string(
        std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now().time_since_epoch()
        ).count()
    );
    if (status_callback_) {
        status_callback_(status_);
    }
}

DeployStatus PayloadDeployer::getStatus() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return status_;
}

bool PayloadDeployer::portOpen(const std::string& host, int port, int timeout_ms) {
    int sock = ::socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) return false;

    struct timeval tv;
    tv.tv_sec = timeout_ms / 1000;
    tv.tv_usec = (timeout_ms % 1000) * 1000;
    setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, (const char*)&tv, sizeof(tv));

    struct sockaddr_in addr = {};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(static_cast<uint16_t>(port));
    inet_pton(AF_INET, host.c_str(), &addr.sin_addr);

    bool result = (::connect(sock, (struct sockaddr*)&addr, sizeof(addr)) == 0);
    closeSocket(sock);
    return result;
}

bool PayloadDeployer::waitPort(const std::string& host, int port, int deadline_ms) {
    auto end = std::chrono::steady_clock::now() + std::chrono::milliseconds(deadline_ms);
    while (std::chrono::steady_clock::now() < end) {
        if (portOpen(host, port, 1000)) return true;
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
    }
    return false;
}

bool PayloadDeployer::deployElf(const std::string& host, const std::string& elf_path, int elf_loader_port) {
    std::ifstream file(elf_path, std::ios::binary | std::ios::ate);
    if (!file.is_open()) return false;

    std::streamsize size = file.tellg();
    file.seekg(0, std::ios::beg);
    std::vector<char> data(size);
    if (!file.read(data.data(), size)) return false;

    int sock = ::socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) return false;

    setNoSigPipe(sock);

    struct timeval tv;
    tv.tv_sec = 30;
    tv.tv_usec = 0;
    setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, (const char*)&tv, sizeof(tv));

    struct sockaddr_in addr = {};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(static_cast<uint16_t>(elf_loader_port));
    inet_pton(AF_INET, host.c_str(), &addr.sin_addr);

    if (::connect(sock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        closeSocket(sock);
        return false;
    }

    int flag = 1;
    setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, (char*)&flag, sizeof(flag));

    ssize_t sent = ::send(sock, data.data(), data.size(), 0);
    closeSocket(sock);
    return sent == static_cast<ssize_t>(data.size());
}

bool PayloadDeployer::sendGbnd(const std::string& host, uint64_t virt, uint64_t phys, uint32_t user) {
    auto packet = buildGbndPacket(user, virt, phys);

    int sock = ::socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) return false;

    setNoSigPipe(sock);

    struct timeval tv;
    tv.tv_sec = 3;
    tv.tv_usec = 0;
    setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, (const char*)&tv, sizeof(tv));

    struct sockaddr_in addr = {};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(6970);
    inet_pton(AF_INET, host.c_str(), &addr.sin_addr);

    if (::connect(sock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        closeSocket(sock);
        return false;
    }

    int flag = 1;
    setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, (char*)&flag, sizeof(flag));

    ::send(sock, (const char*)packet.data(), packet.size(), 0);
    closeSocket(sock);
    return true;
}

void PayloadDeployer::processKlogLine(const std::string& line) {
    fprintf(stderr, "[klog] %s\n", line.c_str());
    fflush(stderr);
    try {
    // Match physical device
    static std::regex re_dev_phys(
        R"(DEVICE_ADDED|DeviceAdded.*?(?:DeviceId|DeviceID|deviceId|deviceID|device id)[:=]\s*0x([0-9a-fA-F]+))",
        std::regex::icase
    );
    static std::regex re_dev_virt(
        R"(DEVICE_ADDED|DeviceAdded|GetUnassignedDeviceInfo.*?(?:DeviceId|DeviceID|deviceId|deviceID|device id)[:=]\s*0x([0-9a-fA-F]+))",
        std::regex::icase
    );
    static std::regex re_owner(
        R"(DEVICE_OWNER_CHANGED.*?deviceId=0x([0-9a-fA-F]+).*?userId=0x([0-9a-fA-F]+))",
        std::regex::icase
    );
    static std::regex re_adopted(R"(VDI active|padHandle)", std::regex::icase);
    static std::regex re_bind_ok(R"(force_bind.*ret=0|MBusBindDeviceWithUserId.*ret=0)", std::regex::icase);

    std::smatch match;

    if (std::regex_search(line, match, re_dev_phys) && match.size() > 1) {
        uint64_t dev = 0;
        try { dev = std::stoull(match[1].str(), nullptr, 16); } catch (...) {}
        if (dev > 0 && (line.find("capabilityBattery:1") != std::string::npos ||
                         line.find("Physical") != std::string::npos)) {
            auto_phys = dev;
            emitStatus("klog", "Physical controller: 0x" + match[1].str());
        }
    }

    if (!auto_sent_gbnd_ && std::regex_search(line, match, re_dev_virt) && match.size() > 1) {
        uint64_t dev = 0;
        try { dev = std::stoull(match[1].str(), nullptr, 16); } catch (...) {}
        bool is_virtual = line.find("capabilityBattery:0") != std::string::npos ||
                          line.find("userId=0xffffffff") != std::string::npos ||
                          line.find("UserId:0xffffffff") != std::string::npos ||
                          line.find("remoteplay") != std::string::npos ||
                          line.find("RemotePlay") != std::string::npos ||
                          line.find("type:4") != std::string::npos ||
                          line.find("VDA candidate") != std::string::npos;
        if (is_virtual && dev > 0) {
            auto_virt = dev;
            auto_sent_gbnd_ = true;
            emitStatus("klog", "Virtual device 0x" + match[1].str() + " detected - sending GBND");

            std::thread([this, dev]() {
                std::this_thread::sleep_for(std::chrono::milliseconds(400));
                sendGbnd(klog_host_, auto_virt, auto_phys, 0);
            }).detach();
        }
    }

    if (std::regex_search(line, match, re_owner)) {
        auto_bound = true;
        emitStatus("klog", "Controller bound: dev=0x" + match[1].str());
    }

    if (std::regex_search(line, re_adopted)) {
        auto_adopted = true;
        emitStatus("klog", "VDI active - GPAD input should work");
    }

    if (std::regex_search(line, re_bind_ok)) {
        auto_bound = true;
        emitStatus("klog", "MBus bind OK");
    }
    } catch (...) {}
}

void PayloadDeployer::processKlogData(const std::string& host, int sock) {
    static std::string buffer;
    char buf[4096];

    while (klog_running_) {
        fd_set fds;
        FD_ZERO(&fds);
        FD_SET(sock, &fds);
        struct timeval tv;
        tv.tv_sec = 0;
        tv.tv_usec = 500000;

        int ret = select(sock + 1, &fds, nullptr, nullptr, &tv);
        if (ret < 0) break;
        if (ret == 0) continue;

        ssize_t n = recv(sock, buf, sizeof(buf) - 1, 0);
        if (n <= 0) break;

        buffer.append(buf, n);
        size_t idx;
        while ((idx = buffer.find('\n')) != std::string::npos) {
            std::string line = buffer.substr(0, idx);
            buffer = buffer.substr(idx + 1);
            if (!line.empty() && line.back() == '\r') line.pop_back();
            processKlogLine(line);
        }
    }
}

void PayloadDeployer::stopKlogWatcher() {
    if (klog_running_) {
        klog_running_ = false;
        if (klog_sock_ >= 0) {
            closeSocket(klog_sock_);
            klog_sock_ = -1;
        }
        if (klog_thread_.joinable()) {
            klog_thread_.join();
        }
    }
}

PayloadDeployer::Result PayloadDeployer::ensurePayloadRunning(const std::string& host, const Options& options) {
    Result result;
    int elf_port = options.elf_loader_port;

    status_.host = host;
    status_callback_ = options.status_callback;

    stopKlogWatcher();
    auto_sent_gbnd_ = false;
    auto_phys = 0;
    auto_virt = 0;
    auto_adopted = false;
    auto_bound = false;

    // Start Klog watcher if auto-bind enabled
    if (options.auto_bind_via_klog) {
        klog_sock_ = ::socket(AF_INET, SOCK_STREAM, 0);
        if (klog_sock_ >= 0) {
            struct sockaddr_in addr = {};
            addr.sin_family = AF_INET;
            addr.sin_port = htons(KLOG_PORT);
            inet_pton(AF_INET, host.c_str(), &addr.sin_addr);

            if (::connect(klog_sock_, (struct sockaddr*)&addr, sizeof(addr)) == 0) {
                klog_host_ = host;
                klog_running_ = true;
                klog_thread_ = std::thread(&PayloadDeployer::processKlogData, this, host, static_cast<int>(klog_sock_));
                emitStatus("klog", "Klog watcher started");
            } else {
                closeSocket(klog_sock_);
                klog_sock_ = -1;
                emitStatus("klog", "Klog port unavailable - continuing without auto-bind");
            }
        }
    }

    bool gp_already = portOpen(host, GPAD_PORT, 1000);

    if (options.force_deploy || !gp_already) {
        if (!portOpen(host, elf_port, 1000)) {
            emitStatus("warn", "ELF loader port " + std::to_string(elf_port) +
                       " is closed - skipping deploy (payload may already be running)");

            if (klog_running_) {
                emitStatus("waiting", "Waiting for direct-VDI adopt (auto-bind)...");
                auto end = std::chrono::steady_clock::now() + std::chrono::milliseconds(12000);
                while (std::chrono::steady_clock::now() < end && !auto_adopted) {
                    std::this_thread::sleep_for(std::chrono::milliseconds(150));
                }
            }

            emitStatus("ready", "Ghostpad ready on " + host + ":" + std::to_string(GPAD_PORT));
            result.ok = true;
            result.skipped = true;
            result.message = "ELF loader port " + std::to_string(elf_port) + " closed - deploy skipped";
            result.adopted = auto_adopted;
            result.bound = auto_bound;
            return result;
        }

        if (options.elf_path.empty()) {
            stopKlogWatcher();
            emitStatus("error", "No payload ELF configured. Choose ghostpad.elf in Settings.");
            result.message = status_.message;
            return result;
        }

        emitStatus("deploying", "Deploying payload to " + host + ":" + std::to_string(elf_port) + "...");
        if (!deployElf(host, options.elf_path, elf_port)) {
            stopKlogWatcher();
            emitStatus("error", "ELF deployment failed");
            result.message = "ELF deployment failed";
            return result;
        }

        emitStatus("waiting", "Payload sent - waiting for GPAD port...");
        if (!waitPort(host, GPAD_PORT, 30000)) {
            stopKlogWatcher();
            emitStatus("error", "Payload deployed but GPAD port 6967 never opened");
            result.message = "Payload deployed but GPAD port 6967 never opened";
            return result;
        }

        if (klog_running_) {
            emitStatus("waiting", "Waiting for direct-VDI adopt (auto-bind)...");
            auto end = std::chrono::steady_clock::now() + std::chrono::milliseconds(12000);
            while (std::chrono::steady_clock::now() < end && !auto_adopted) {
                std::this_thread::sleep_for(std::chrono::milliseconds(150));
            }
        }
    } else {
        emitStatus("ready", "Payload already running on " + host + " - reusing existing session");
    }

    emitStatus("ready", "Ghostpad ready on " + host + ":" + std::to_string(GPAD_PORT));
    result.ok = true;
    result.message = status_.message;
    result.adopted = auto_adopted;
    result.bound = auto_bound;
    return result;
}

} // namespace ghostpad

File: deprecated/Ghostpad-app/electron/payload-deploy.js
const net = require("net");
const fs = require("fs");

const GPAD_PORT = 6967;
const ELF_LOAD_PORT = 9021;
const CTRL_PORT = 6970;
const KLOG_PORT = 3434;

const RE_DEV_PHYS =
  /DEVICE_ADDED|DeviceAdded.*?(?:DeviceId|DeviceID|deviceId|deviceID|device id)[:=]\s*0x([0-9a-f]+)/i;
const RE_DEV_VIRT =
  /DEVICE_ADDED|DeviceAdded|GetUnassignedDeviceInfo.*?(?:DeviceId|DeviceID|deviceId|deviceID|device id)[:=]\s*0x([0-9a-f]+)/i;
const RE_OWNER =
  /DEVICE_OWNER_CHANGED.*?deviceId=0x([0-9a-f]+).*?userId=0x([0-9a-f]+)/i;
const RE_ADOPTED = /VDI active|padHandle/i;
const RE_BIND_OK = /force_bind.*ret=0|MBusBindDeviceWithUserId.*ret=0/i;

let activeWatcher = null;
let lastDeployStatus = {
  phase: "idle",
  message: "Ready",
  host: null,
  at: null,
};

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function setDeployStatus(phase, message, host) {
  lastDeployStatus = {
    phase,
    message,
    host: host ?? lastDeployStatus.host,
    at: new Date().toISOString(),
  };
  return lastDeployStatus;
}

function portOpen(host, port, timeoutMs = 1000) {
  return new Promise((resolve) => {
    const socket = new net.Socket();
    let settled = false;
    const finish = (result) => {
      if (settled) return;
      settled = true;
      try {
        socket.destroy();
      } catch (_) {
        /* ignore */
      }
      resolve(result);
    };
    socket.setTimeout(timeoutMs);
    socket.once("connect", () => finish(true));
    socket.once("timeout", () => finish(false));
    socket.once("error", () => finish(false));
    socket.connect(port, host);
  });
}

async function waitPort(host, port, deadlineMs) {
  const end = Date.now() + deadlineMs;
  while (Date.now() < end) {
    if (await portOpen(host, port, 1000)) {
      return true;
    }
    await sleep(500);
  }
  return false;
}

async function deployElf(host, elfPath, elfLoaderPort) {
  if (!elfPath || !fs.existsSync(elfPath)) {
    return { ok: false, message: `ELF missing at ${elfPath || "(not set)"}` };
  }

  const port = elfLoaderPort || ELF_LOAD_PORT;

  let data;
  try {
    data = fs.readFileSync(elfPath);
  } catch (err) {
    return { ok: false, message: `Read ELF: ${err.message}` };
  }

  return new Promise((resolve) => {
    const socket = new net.Socket();
    socket.setTimeout(30000);
    socket.once("error", (err) =>
      resolve({
        ok: false,
        message: `Send to ${host}:${port}: ${err.message}`,
      })
    );
    socket.connect(port, host, () => {
      socket.write(data, (err) => {
        socket.end();
        if (err) {
          resolve({ ok: false, message: err.message });
        } else {
          resolve({ ok: true, message: "Payload sent" });
        }
      });
    });
  });
}

function sendGbnd(host, virt, phys, user = 0) {
  // 4 (magic) + 4 (userId u32) + 8 (virt u64) + 8 (phys u64) = 24 bytes
  const buf = Buffer.alloc(24);
  buf.write("GBND", 0);
  buf.writeUInt32LE(user >>> 0, 4);
  buf.writeBigUInt64LE(BigInt(virt), 8);
  buf.writeBigUInt64LE(BigInt(phys), 16);

  return new Promise((resolve) => {
    const socket = new net.Socket();
    socket.setTimeout(3000);
    socket.once("error", (err) => resolve({ ok: false, message: err.message }));
    socket.connect(CTRL_PORT, host, () => {
      socket.write(buf, () => {
        socket.end();
        resolve({ ok: true });
      });
    });
  });
}

class KlogWatcher {
  constructor(host, onStatus) {
    this.host = host;
    this.onStatus = onStatus || (() => {});
    this.autoPhys = 0;
    this.autoVirt = 0;
    this.autoBound = false;
    this.autoAdopted = false;
    this.autoSentGbnd = false;
    this.socket = null;
    this.buffer = "";
  }

  start() {
    return new Promise((resolve) => {
      const socket = new net.Socket();
      this.socket = socket;
      socket.setTimeout(500);
      socket.once("error", () => resolve(false));
      socket.connect(KLOG_PORT, this.host, () => resolve(true));
      socket.on("data", (chunk) => {
        this.processChunk(chunk.toString("utf8", "replace"));
      });
      socket.on("timeout", () => {
        /* keep listening */
      });
    });
  }

  processChunk(text) {
    this.buffer += text;
    let idx = this.buffer.indexOf("\n");
    while (idx >= 0) {
      const line = this.buffer.slice(0, idx);
      this.buffer = this.buffer.slice(idx + 1);
      this.processLine(line);
      idx = this.buffer.indexOf("\n");
    }
  }

  processLine(line) {
    let match = RE_DEV_PHYS.exec(line);
    if (match) {
      const dev = parseInt(match[1], 16);
      if (dev && (line.includes("capabilityBattery:1") || line.includes("Physical"))) {
        this.autoPhys = dev;
        this.onStatus(`Physical controller: 0x${dev.toString(16)}`);
      }
    }

    match = RE_DEV_VIRT.exec(line);
    if (match && !this.autoSentGbnd) {
      const dev = parseInt(match[1], 16);
      const isVirtual =
        line.includes("capabilityBattery:0") ||
        line.includes("userId=0xffffffff") ||
        line.includes("UserId:0xffffffff") ||
        line.includes("remoteplay") ||
        line.includes("RemotePlay") ||
        line.includes("type:4") ||
        line.includes("VDA candidate");
      if (isVirtual && dev) {
        this.autoVirt = dev;
        this.autoSentGbnd = true;
        this.onStatus(
          `Virtual device 0x${this.autoVirt.toString(16)} detected — sending GBND`
        );
        setTimeout(() => {
          try {
            sendGbnd(this.host, this.autoVirt, this.autoPhys, 0).catch(() => {});
          } catch (e) {
            console.error("sendGbnd threw:", e);
          }
        }, 400);
      }
    }

    match = RE_OWNER.exec(line);
    if (match) {
      this.autoBound = true;
      this.onStatus(
        `Controller bound: dev=0x${parseInt(match[1], 16).toString(16)}`
      );
    }

    if (RE_ADOPTED.test(line)) {
      this.autoAdopted = true;
      this.onStatus("VDI active — GPAD input should work");
    }
    if (RE_BIND_OK.test(line)) {
      this.autoBound = true;
      this.onStatus("MBus bind OK");
    }
  }

  stop() {
    if (this.socket) {
      try {
        this.socket.destroy();
      } catch (_) {
        /* ignore */
      }
      this.socket = null;
    }
  }
}

function stopKlogWatcher() {
  if (activeWatcher) {
    activeWatcher.stop();
    activeWatcher = null;
  }
}

async function ensurePayloadRunning(host, options = {}) {
  const {
    elfPath,
    forceDeploy = false,
    autoBindViaKlog = true,
    onStatus,
    elfLoaderPort,
  } = options;

  const port = elfLoaderPort || ELF_LOAD_PORT;

  const emit = (phase, message) => {
    const status = setDeployStatus(phase, message, host);
    if (onStatus) onStatus(status);
    return status;
  };

  stopKlogWatcher();

  let watcher = null;
  if (autoBindViaKlog) {
    watcher = new KlogWatcher(host, (message) => emit("klog", message));
    const klogOk = await watcher.start();
    if (!klogOk) {
      emit("klog", "Klog port unavailable — continuing without auto-bind");
      watcher.stop();
      watcher = null;
    } else {
      activeWatcher = watcher;
    }
  }

  const gpAlready = await portOpen(host, GPAD_PORT, 1000);

  if (forceDeploy || !gpAlready) {
    if (!(await portOpen(host, port, 1000))) {
      emit(
        "warn",
        `ELF loader port ${port} is closed — skipping deploy (payload may already be running)`
      );
      if (watcher) {
        emit("waiting", "Waiting for direct-VDI adopt (auto-bind)...");
        const end = Date.now() + 12000;
        while (Date.now() < end && !watcher.autoAdopted) {
          await sleep(150);
        }
      }
      emit("ready", `Ghostpad ready on ${host}:${GPAD_PORT}`);
      return {
        ok: true,
        skipped: true,
        message: `ELF loader port ${port} closed — deploy skipped`,
      };
    }

    if (!elfPath) {
      if (watcher) watcher.stop();
      activeWatcher = null;
      emit(
        "error",
        "No payload ELF configured. Choose ghostpad.elf in Settings."
      );
      return { ok: false, message: lastDeployStatus.message };
    }

    emit("deploying", `Deploying payload to ${host}:${port}...`);
    const deployResult = await deployElf(host, elfPath, port);
    if (!deployResult.ok) {
      if (watcher) watcher.stop();
      activeWatcher = null;
      emit("error", deployResult.message);
      return { ok: false, message: deployResult.message };
    }

    emit("waiting", "Payload sent — waiting for GPAD port...");
    if (!(await waitPort(host, GPAD_PORT, 30000))) {
      if (watcher) watcher.stop();
      activeWatcher = null;
      emit("error", "Payload deployed but GPAD port 6967 never opened");
      return { ok: false, message: lastDeployStatus.message };
    }

    if (watcher) {
      emit("waiting", "Waiting for direct-VDI adopt (auto-bind)...");
      const end = Date.now() + 12000;
      while (Date.now() < end && !watcher.autoAdopted) {
        await sleep(150);
      }
    }
  } else {
    emit("ready", `Payload already running on ${host} — reusing existing session`);
  }

  emit("ready", `Ghostpad ready on ${host}:${GPAD_PORT}`);
  return {
    ok: true,
    message: lastDeployStatus.message,
    adopted: watcher?.autoAdopted ?? false,
    bound: watcher?.autoBound ?? false,
  };
}

function getDeployStatus() {
  return lastDeployStatus;
}

module.exports = {
  GPAD_PORT,
  ELF_LOAD_PORT,
  ensurePayloadRunning,
  deployElf,
  getDeployStatus,
  stopKlogWatcher,
  portOpen,
};
